xarray/coding/variables.py

```diff
@@ -349,46 +361,115 @@ def encode(self, variable: Variable, name: T_Name = None):
+        try:
+            # user provided the on-disk signed fill
+            if hasattr(fill_value, "item"):
+                # if numpy type, convert to python native integer to determine overflow
+                # otherwise numpy unsigned ints will silently cast to the signed counterpart
+                fill_value = fill_value.item()
+            new_fill = signed_dtype.type(fill_value)
+        except OverflowError:
+            warnings.warn(
+                f"variable {name!r} will be stored as signed integers "
+                f"but _FillValue attribute can't be represented as a "
+                f"signed integer.",
+                SerializationWarning,
+                stacklevel=3,
+            )
+            # user provided the in-memory unsigned fill, convert to signed type
+            unsigned_dtype = np.dtype(f"u{signed_dtype.itemsize}")
+            # use view here to prevent OverflowError
+            new_fill = (
+                np.array(fill_value, dtype=unsigned_dtype).view(signed_dtype).item()
+            )
+        return new_fill
+
     def decode(self, variable: Variable, name: T_Name = None):
         raw_fill_dict, encoded_fill_values = _check_fill_values(
             variable.attrs, name, variable.dtype
         )
+        if "_Unsigned" not in variable.attrs and not raw_fill_dict:
+            return variable
 
-        if raw_fill_dict:
-            dims, data, attrs, encoding = unpack_for_decoding(variable)
-            [
-                safe_setitem(encoding, attr, value, name=name)
-                for attr, value in raw_fill_dict.items()
-            ]
-
-            if encoded_fill_values:
-                # special case DateTime to properly handle NaT
-                dtype: np.typing.DTypeLike
-                decoded_fill_value: Any
-                if _is_time_like(attrs.get("units")) and data.dtype.kind in "iu":
-                    dtype, decoded_fill_value = np.int64, np.iinfo(np.int64).min
-                else:
-                    if "scale_factor" not in attrs and "add_offset" not in attrs:
-                        dtype, decoded_fill_value = dtypes.maybe_promote(data.dtype)
-                    else:
-                        dtype, decoded_fill_value = (
-                            _choose_float_dtype(data.dtype, attrs),
-                            np.nan,
-                        )
+        dims, data, attrs, encoding = unpack_for_decoding(variable)
+
+        # dims, data, attrs, encoding = unpack_for_decoding(variable)
+        # Even if _Unsigned is use, retain on-disk _FillValue
+        [
+            safe_setitem(encoding, attr, value, name=name)
+            for attr, value in raw_fill_dict.items()
+        ]
 
-                transform = partial(
-                    _apply_mask,
-                    encoded_fill_values=encoded_fill_values,
-                    decoded_fill_value=decoded_fill_value,
-                    dtype=dtype,
+        if "_Unsigned" in attrs:
```

Comment:
can we factor this out to its own function please

Reply:
That's why it is a draft and I said "At the time of writing this code needs lots of cleanup or at least I hope it can because it is hard to follow in my opinion".

Reply:
Very basic refactoring done for this specific if block. But it is leading me to some questions that I'll comment inline...